File: app/course-enrollment/app/rabbitmq/rabbitMQ_service.py
import asyncio
import json
import uuid
import aio_pika
from aio_pika import Message, connect_robust
from aio_pika.abc import AbstractChannel, AbstractConnection
from app.config import RABBITMQ_URL

connection: AbstractConnection = None
channel: AbstractChannel = None

async def initialize_rabbitmq():
    global connection, channel
    connection = await connect_robust(RABBITMQ_URL)
    channel = await connection.channel()

# Call once on app startup
asyncio.create_task(initialize_rabbitmq())

async def publish_message_with_reply(queue_name: str, payload: dict, timeout: int = 5):
    if not channel:
        raise RuntimeError("RabbitMQ channel not initialized")

    correlation_id = str(uuid.uuid4())

    # Create a temporary exclusive reply queue
    reply_queue = await channel.declare_queue(exclusive=True)

    future = asyncio.get_event_loop().create_future()

    async def on_message(message: aio_pika.IncomingMessage):
        if message.correlation_id == correlation_id:
            future.set_result(json.loads(message.body.decode()))
            await message.ack()

    await reply_queue.consume(on_message, no_ack=False)

    message = Message(
        body=json.dumps(payload).encode(),
        correlation_id=correlation_id,
        reply_to=reply_queue.name,
        content_type='application/json'
    )

    await channel.default_exchange.publish(
        message, routing_key=queue_name
    )

    try:
        return await asyncio.wait_for(future, timeout=timeout)
    except asyncio.TimeoutError:
        raise TimeoutError("Request timed out")

File: app/api-gateway/services/rabbitmqService.js
// import amqp from "amqplib";
import { connectRabbitMQ } from "../config/rabbitMQ.js";
import { v4 as uuidv4 } from 'uuid';

let channel = null;

const initializeRabbitMQ = async () => {
  channel = await connectRabbitMQ();
};

initializeRabbitMQ();

export const publishMessageWithReply = async (queue, message) => {
  if (!channel) {
    console.error("RabbitMQ channel not initialized");
    return;
  }

  const correlationId = uuidv4();

  const replyQueue = await channel.assertQueue('', { exclusive: true });

  return new Promise((resolve, reject) => {

    const timeout = setTimeout(() => {
      reject(new Error("Request timed out"));
    }, 5000);

    channel.consume(
      replyQueue.queue,
      (msg) => {
        if (msg.properties.correlationId === correlationId) {
          clearTimeout(timeout); // ✅ clear timeout if response is received
          const response = JSON.parse(msg.content.toString());
          resolve(response);
        }
      },
      { noAck: true }
    );

    channel.sendToQueue(queue, Buffer.from(JSON.stringify(message)), {
      correlationId: correlationId,
      replyTo: replyQueue.queue,
    });
  });
};